### ISSUE TITLE
Diff output discards assertion details

### ISSUE THREAD

**Issue by bajtos** [mochajs/mocha#1065]
When a test failure is printed using the diff output, assertion details are discarded by mocha. This makes it more difficult to understand why the test failed without inspecting test source code. It is also inconsistent with the default (non-diff) algorithm.

```js
expect(undefined, 'first name').to.equal('expected');
/* result:
AssertionError: first name: expected undefined to equal 'expected'
*/

expect('actual', 'surname').to.equal('expected');
/* result:
actual expected
"expectedactual"
*/
```

**Update**

Another bit of information that is stripped from the output is the stack trace, as pointed out in the mailing list ([link](https://groups.google.com/forum/#!msg/mochajs/2ibth8zs7vY/lEffhy7Mz8MJ)).

```js
expect(false).to.equal(true);
/* result:
  + expected - actual

  +true
  -false
*/

assert.equal(false, true);
/* result:
AssertionError: true == false
 at Context.<anonymous> (/private/tmp/asd/test.js:5:12)
 at Test.Runnable.run (/usr/local/lib/node_modules/mocha/lib/runnable.js:211:32)
 [etc.]
*/
```

**Comment by boneskull** [comment]
closing as old

**Comment by bajtos** [comment]
@boneskull I believe the problem was still not fixed :( Are you willing to accept a pull request from me that would fix the issue? The issue is old because I was waiting for input from Mocha maintainers on what's the acceptable solution, since my initial attempt to fix this issue via #949 was rejected.

**Comment by clns** [comment]
This is still a pretty big problem (with the default reporter at least), especially when using the [chai](http://chaijs.com/) library. For something like this:

```js
expect(1).to.equal(2)
```

You get:
![image](https://cloud.githubusercontent.com/assets/687922/14062071/2d451412-f39b-11e5-81f1-c7a0e84fdcb0.png)

But if you use a description (e.g. `expect(1, 'My description').to.equal(2)`), you get:

![image](https://cloud.githubusercontent.com/assets/687922/14062082/7fa5a582-f39b-11e5-8421-e5d256ee5e1e.png)

And it gets even worse when using `.not` (e.g. `expect(1, 'My description').to.not.equal(1)`) since there's no diff to output so you will see just the description:
![image](https://cloud.githubusercontent.com/assets/687922/14062092/a332d9fc-f39b-11e5-94b0-066e4e046668.png)

This happens because chai sets the `AssertionError.message` to **My description: expected 1 to not equal 1** and mocha deletes everything except what is before `: expected` in [lib/reporters/base.js](https://github.com/mochajs/mocha/blob/e0a43d2f5a39a55e5fe62b4c8734767b8c7fc86e/lib/reporters/base.js#L208).
